### PR TITLE
Update ClientGateway&EthereumNodeService to use selected network

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/Cross-layer/App.swift [hunk 1]
```diff
@@ -53,5 +51,23 @@ class App {
...
+
+        for name in updateNotifications {
+            NotificationCenter.default.addObserver(self,
+                                           selector: #selector(bindNetworkInfo),
+                                           name: name,
+                                           object: nil)
+        }
+    }
+
+    @objc func bindNetworkInfo() {
```
Comment (sche) [inline Multisig/Cross-layer/App.swift]: could be private

File: Multisig/Data/Services/Safe Client Gateway Service/ConfirmTransactionRequest.swift [hunk 2]
```diff
@@ -11,8 +11,10 @@ import Foundation
 struct ConfirmTransactionRequest: JSONRequest {
     var safeTxHash: String
     var signedSafeTxHash: String
+    let network: Int
+    
     var httpMethod: String { "POST" }
-    var urlPath: String { "/v1/transactions/\(safeTxHash)/confirmations" }
+    var urlPath: String { "/\(network)/v1/transactions/\(safeTxHash)/confirmations" }
```
Comment (rmeissner) [inline Multisig/Data/Services/Safe Client Gateway Service/ConfirmTransactionRequest.swift]: The naming seems inconsistent with the other occurrences where it is called `chainId` ;)

File: Multisig/Cross-layer/App.swift (hunk 1 above)
Comment (MouazAlzahabi) [reply Multisig/Cross-layer/App.swift]: No, it's used at AppDelegate

File: Multisig/Cross-layer/App.swift [hunk 3]
```diff
@@ -53,5 +51,23 @@ class App {
...
+        if let network = try? Safe.getSelected()?.network {
+            clientGatewayService.bindNetwork(Int(network.chainId))
```
Comment (sche) [inline Multisig/Cross-layer/App.swift]: This approach may break monitoring functionality. Right now we monitor pending WalletConnect transactions, but in future we may monitor more things - e.g. if we add tx execution functionality we might monitor tx status.
On the other hand we can add optional `network` parameter in future.

File: Multisig/Cross-layer/App.swift (hunk 1 above)
Comment (sche) [reply Multisig/Cross-layer/App.swift]: or can't because you use it in AppDelegate?

File: Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift [hunk 4]
```diff
@@ -35,8 +34,8 @@ struct BalancesRequest: JSONRequest {
 }
 
 extension BalancesRequest {
-    init(_ address: Address) {
-        self.init(safeAddress: address.checksummed, fiat: AppSettings.selectedFiatCode)
+    init(_ address: Address, chainId: Int) {
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift]: As a next refactoring step (our of scope of this PR) I would look into passing everywhere Safe object, that already has it's network, so probably we can remove network binding logic

File: Multisig/Cross-layer/App.swift (hunk 1 above)
Comment (DmitryBespalov) [inline Multisig/Cross-layer/App.swift]: 👍

File: Multisig/Data/Services/Safe Client Gateway Service/FiatCodesRequest.swift [hunk 5]
```diff
@@ -9,19 +9,19 @@
...
-    var httpMethod: String {
-        "GET"
-    }
+    let chainId: Int
+    
+    var httpMethod: String { "GET" }
 
     var urlPath: String {
-        "/v1/balances/supported-fiat-codes"
+        "/\(chainId)/v1/balances/supported-fiat-codes"
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/FiatCodesRequest.swift]: As I can see here https://github.com/gnosis/safe-client-gateway/blob/b152f3e3d27c85efeb96c9baf8b43c7f5f8a6d93/src/routes/balances.rs#L62 there is no need in chain id here

File: Multisig/Data/Services/Safe Client Gateway Service/TransactionDetailsRequest.swift [hunk 6]
```diff
@@ -10,23 +10,26 @@ import Foundation
...
 
 extension TransactionDetailsRequest {
-    init(safeTxHash: Data) {
+    init(safeTxHash: Data, chainId: Int) {
         id = safeTxHash.toHexStringWithPrefix()
+        self.chainId = chainId
     }
 }
 
 extension SafeClientGatewayService {
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/TransactionDetailsRequest.swift:27]: @MouazAlzahabi did you run integration tests? I think client gateway tests will break now

File: Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift (hunk 4 above)
Comment (DmitryBespalov) [reply Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift]: @sche hm, maybe that's not a good one, because this will couple the request to the database model, I think the current approach is good

File: Multisig/Data/Services/Safe Client Gateway Service/ProposeTransactionRequest.swift [hunk 7]
```diff
@@ -67,13 +68,17 @@ extension SafeClientGatewayService {
         return asyncExecute(request: ProposeTransactionRequest(safe: transaction.safe!,
                                                                sender: sender,
                                                                signature: signature,
-                                                               transaction: transaction),
+                                                               transaction: transaction,
+                                                               chainId: chainId),
                             completion: completion)
     }
 
     func proposeTransaction(transaction: Transaction, sender: AddressString, signature: String) throws {
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/ProposeTransactionRequest.swift]: e.g. here transaction should already have attached safe, so chainId could be fetched as `transaction.safe.network!.id`

File: Multisig/Data/Services/Safe Client Gateway Service/ConfirmTransactionRequest.swift (hunk 2 above)
Comment (sche) [reply Multisig/Data/Services/Safe Client Gateway Service/ConfirmTransactionRequest.swift]: I would rename in other places into `network` as we use this name in our model

Comment (MouazAlzahabi) [reply Multisig/Data/Services/Safe Client Gateway Service/ConfirmTransactionRequest.swift]: Yes, by mistake :)

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 8]
```diff
@@ -538,6 +538,28 @@ extension SCGModels {
...
+            case nativeCurrency
+            case theme
+        }
+
+        init(from decoder: Decoder) throws {
+            let values = try decoder.container(keyedBy: CodingKeys.self)
+            var string = try! values.decode(String.self, forKey: .chainId)
+            chainId = Int(string) ?? 0
+            chainName = try values.decode(String.self, forKey: .chainName)
+            string = try! values.decode(String.self, forKey: .rpcUrl)
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: why do you mix `try!` and `try`?

File: Multisig/Data/Services/Safe Client Gateway Service/SafeClientGatewayService.swift [hunk 9]
```diff
@@ -10,6 +10,7 @@ import Foundation
 
 class SafeClientGatewayService {
     var url: URL
+    var chainId: Int! = 1
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/SafeClientGatewayService.swift]: we have a enum: Network.ChainID.mainEthereumNetwork

File: Multisig/Logic/Models/Network.swift [hunk 10]
```diff
@@ -91,7 +91,7 @@ extension Network {
     static func create(_ networkInfo: SCGModels.Network) throws -> Network {
         try Network.create(chainId: networkInfo.chainId,
                      chainName: networkInfo.chainName,
-                     rpcUrl: networkInfo.authenticatedRpcUrl,
+                     rpcUrl: networkInfo.rpcUrl,
```
Comment (sche) [inline Multisig/Logic/Models/Network.swift:92]: why did you change it here?

File: Multisig/Logic/NetworkManager.swift [hunk 11]
```diff
@@ -13,12 +13,12 @@ class NetworkManager {
...
                 switch result {
-                case .success(let networks):
-                    for network in networks {
+                case .success(let data):
+                    for network in data.results {
                         Network.updateIfExist(network)
                     }
                     NotificationCenter.default.post(name: .networkInfoChanged, object: nil)
-                case .failure(_):
+                case .failure(let error):
```
Comment (sche) [inline Multisig/Logic/NetworkManager.swift]: no need if we ignore

File: Multisig/Logic/Models/Network.swift (hunk 10 above)
Comment (MouazAlzahabi) [reply Multisig/Logic/Models/Network.swift:92]: Because by the old why we store Infura key in the database as part of rpc url

File: Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift [hunk 12]
```diff
@@ -9,17 +9,17 @@
...
 
 struct BalancesRequest: JSONRequest {
-    var safeAddress: String
-    var fiat: String
-    var isTrusted: Bool?
-    var isExcludeSpam: Bool?
+    private let safeAddress: String
+    private let networkId: Int
+    private let fiat: String
+    #warning("Check if we need these two")
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/BalancesRequest.swift]: we don't use them because their default values are ok for mobile

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 13]
```diff
@@ -538,6 +538,42 @@ extension SCGModels {
...
+            self.chainName = chainName
+            self.rpcUrl = rpcUrl
+            self.blockExplorerUrl = blockExplorerUrl
+            self.nativeCurrency = nativeCurrency
+            self.theme = theme
+        }
+
+        init(from decoder: Decoder) throws {
+            let values = try decoder.container(keyedBy: CodingKeys.self)
+            var string = try! values.decode(String.self, forKey: .chainId)
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: I would not use forced try and forced unwrapping everywhere here

Comment (DmitryBespalov) [reply Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: I would throw

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 14]
```diff
@@ -538,6 +538,42 @@ extension SCGModels {
...
+        enum CodingKeys: String, CodingKey {
+            case chainId
+            case chainName
+            case rpcUrl
+            case blockExplorerUrl
+            case nativeCurrency
+            case theme
+        }
+
+        init(chainId: Int,
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: Instead of implementing these here, I would implement extension on URL that would make it Decodable

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift (hunk 13 above)
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: I think if you'd define the value type as UInt256String, it will accept the stringified integer that can be used in other places, so we don't need to implement decodable here

File: Multisig/UI/WalletConnect/WCRequestsHandler.swift [hunk 15]
```diff
@@ -99,7 +102,9 @@ class WCRequestsHandler: RequestHandler {
             }
         } else {
             do {
-                let result = try App.shared.nodeService.rawCall(payload: request.jsonString)
+                #warning("Make sure if this always works @Andrey")
+                let rpcURL = try! Safe.getSelected()!.network!.authenticatedRpcUrl
```
Comment (DmitryBespalov) [inline Multisig/UI/WalletConnect/WCRequestsHandler.swift:106]: @sche did you see it?

Comment (MouazAlzahabi) [reply Multisig/UI/WalletConnect/WCRequestsHandler.swift:106]: He asked me to put this

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 16]
```diff
@@ -532,12 +532,16 @@ extension SCGModels {
 
 
     struct Network: Decodable {
-        let chainId: Int
+        let chainId: UInt256String
```
Comment (sche) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift:535]: 👍🏻

**Review by sche — APPROVED** [review]

File: bin/test.sh [hunk 17]
```diff
@@ -1,4 +1,5 @@
 #! /usr/bin/env bash
+set -e +o pipefail
```
Comment (sche) [inline bin/test.sh:2]: @DmitryBespalov could it be a reason why unit tests can not run on Jenkins?

Comment (DmitryBespalov) [reply bin/test.sh:2]: I think, this is the reason:
Screenshot 2021-07-01 at 11 45 26

**Review by DmitryBespalov — APPROVED** [review]
